### PR TITLE
Sum dataset

### PR DESCRIPTION
**Pull request by LuisFSegalla**
- [x] Depends on #1 

Added the sum of all N channels connected to the detector as a dataset.

### CODE REVIEW

File: cpp/data/frameProcessor/src/XspressProcessPlugin.cpp [hunk 1]
```diff
@@ -496,88 +489,115 @@ void XspressProcessPlugin::process_frame(boost::shared_ptr <Frame> frame)
         // This must be offset according to the rank and number of processes
         uint32_t push_frame_id = ((frame_id / frames_per_block_) * concurrent_processes_) + concurrent_rank_;
         FrameMetaData mca_metadata(push_frame_id, ss.str(), raw_32bit, "", mca_dims);
-        boost::shared_ptr<Frame> mca_frame(new DataBlockFrame(mca_metadata, memory_ptrs_[index]->current_byte_size()));
-        memcpy(mca_frame->get_data_ptr(), memory_ptrs_[index]->get_data_ptr(), memory_ptrs_[index]->current_byte_size());
+        boost::shared_ptr<Frame> mca_frame(new DataBlockFrame(mca_metadata, memory_ptrs_[index]->size()));
+        memcpy(mca_frame->get_data_ptr(), memory_ptrs_[index]->get_data_ptr(), memory_ptrs_[index]->size());
         // Set the chunking size
-        mca_frame->set_outer_chunk_size((int)memory_ptrs_[index]->frames());
+        mca_frame->set_outer_chunk_size(frames_per_block_);
```
Comment (GDYendell) [inline cpp/data/frameProcessor/src/XspressProcessPlugin.cpp:518]: This isn't formatting.

**Comment by GDYendell** [comment]
I am not sure about the reformatting. I think it best to just stick with the current formatting unless we implement some tooling for format checking in CI and ideally auto formatting, because otherwise it will have to manually reviewed every time.

File: cpp/data/frameProcessor/src/XspressProcessPlugin.cpp (hunk 1 above)
Comment (LuisFSegalla) [reply cpp/data/frameProcessor/src/XspressProcessPlugin.cpp:518]: So, I think this is actually an error from git. I checked this part of the code and this is an if/else inside this [loop](https://github.com/DiamondLightSource/xspress-detector/blob/main/cpp/data/frameProcessor/src/XspressProcessPlugin.cpp#L442) and it seems like git got confused with all the formatting changes and is reporting that what's inside the **else** statement got changed, when in reality this is just the **if** statement that got moved around when everything was formatted.

**Comment by GDYendell** [comment]
OK that looks better without the formatting of the existing code, but could you also match the formatting of the existing code with the new code, e.g. indentation does not match.

File: cpp/data/frameProcessor/src/XspressProcessPlugin.cpp [hunk 2]
```diff
@@ -345,6 +347,26 @@ void XspressProcessPlugin::setup_memory_allocation()
     free(inp_est_memblock_);
   }
   inp_est_memblock_ = malloc(sizeof(double) * this->frames_per_block_ * num_channels_);
+
+  if (sum_memblock_)
+  {
+    free(sum_memblock_);
+  }
+  sum_memblock_ = malloc(sizeof(double) * this->frames_per_block_);
+  }
```
Comment (GDYendell) [inline cpp/data/frameProcessor/src/XspressProcessPlugin.cpp:356]: This bracket being indented really confused me

Comment (LuisFSegalla) [reply cpp/data/frameProcessor/src/XspressProcessPlugin.cpp:356]: My bad, I did the indent in a block and got the bracket by accident ):

File: cpp/data/frameProcessor/src/XspressProcessPlugin.cpp [hunk 3]
```diff
@@ -345,6 +347,26 @@ void XspressProcessPlugin::setup_memory_allocation()
...
+  const uint32_t *data = static_cast<const uint32_t *>(frame->get_image_ptr());
+  size_t elements_count = frame->get_image_size() / sizeof(uint32_t);
+  for (size_t pixel_index = 0; pixel_index < elements_count; pixel_index++)
+  {
+    sum_value += data[pixel_index];
+  }
+  double *sum_ptr = (double *)sum_memblock_;
+  sum_ptr += num_scalars_recorded_;
+  memcpy(sum_ptr, &sum_value, sizeof(double));
 }
```
Comment (GDYendell) [inline cpp/data/frameProcessor/src/XspressProcessPlugin.cpp:370]: This needs a docstring explaining what it is doing. Is it summing `num_scalars_recorded_` values and then inserting that at the offset that would be the end of that block in the data array, leaving the rest blank? Why does it do this?

Comment (LuisFSegalla) [reply cpp/data/frameProcessor/src/XspressProcessPlugin.cpp:370]: So, **sum_ptr** is an array with the size equals to the number of frames in your block. What I was trying to do in that part of the code is use the **num_scalars_recorded_** variable to shift the position inside the array to which we were going to write the summed up data. So each frame is received and summed and then added to the corresponding position in the **sum_ptr** array.

Comment (GDYendell) [reply cpp/data/frameProcessor/src/XspressProcessPlugin.cpp:370]: That would be good. I am still not sure I understand why you are shifting `sum_ptr` by `num_scalars_recorded` and not just `1`. Are you intentionally leaving gaps of unset values in `sum_memblock_` and if so why?

Comment (GDYendell) [reply cpp/data/frameProcessor/src/XspressProcessPlugin.cpp:370]: Oh sorry! I am being silly. It resets to the start of the memory block each time this is called and then you set it to where you have gotten to each time. The first time I read this I was thinking it increments by `num_scalars_recorded_` every time.

OK that's fine so just add a docstring saying it calculates the sum for the frame and inserts the result into the next free position based on the number of scalars recorded so far, or something like that.

Comment (LuisFSegalla) [reply cpp/data/frameProcessor/src/XspressProcessPlugin.cpp:384]: Cool!
I added some explanation of what the method does!

File: cpp/data/frameProcessor/src/XspressProcessPlugin.cpp [hunk 4]
```diff
@@ -345,6 +347,40 @@ void XspressProcessPlugin::setup_memory_allocation()
...
+/**
+ * Calculate the sum for a frame
+ *
+ * This method will receive a frame, and sum each element of data on it.
+ * The data is stored in the `sum_memblock_` array (allocated in the `setup_memory_allocation method`)
+ * and we iterate to by casting a pointer (`sum_ptr`) to the beginning of the array and moving it 
+ * by `num_scalars_recorded_` each time this method is called again.
+ * The 'num_scalars_recorded_' variable is set back to zero each time the frame arrays are flushed.
+ * i.e.: when num_scalars_recorded_ is 0 we move the sum_ptr pointer to index 0,
+ * when num_scalars_recorded_ is 1 we move the sum_ptr to index 1 based
```
Comment (GDYendell) [inline cpp/data/frameProcessor/src/XspressProcessPlugin.cpp:368]: I think my questions have made you overthink your explanation! I think this is all you need to say:

```suggestion
 * and each time this method is called it stores the result to this array at the offset `num_scalars_recorded_`, i.e.: when
 * `num_scalars_recorded_` is 0 we store the sum to index 0, when `num_scalars_recorded_` is 1 we store the sum to index 1.
 * The 'num_scalars_recorded_' variable is set back to zero each time the frame arrays are flushed.
```

**Review by GDYendell — APPROVED** [review]

**Pull request merged**